### PR TITLE
 PytestFramework: Add initial test framework

### PR DESCRIPTION
**Pull request by mitzkia**
**Current functionality**

* Because it was too hard to review the previous version of the framework, we decied to minimalize it as much as possible.
* The idea was to be able to write a "smallest" testcase (in used features).

## How to use the framework

1, Dependencies:
 * The test-framework was built on top of Python2 and Python3 support
 * The root directory of test-framework contains a requirements.txt file which contains framework dependencies.

2, Usage with make:
 * Here you cannot use custom pytest commands (yet).
 * This mode has only 2 hardcoded commands:
   * `$ make pytest-check` This command runs all functional tests under "functional_test_examples/" dir without verbose console log
   * `$ make pytest-self-check` This command runs all self tests for test-framework

3, Usage without make:
 * Here you can use your favorite pytest commands
 * Here you must define your syslog-ng install path after: "--installdir" option
 * Example1: A standard start for all functional tests with verbose output on console
   * `$ cd tests/pytest_framework`
   * `$ python3 -m pytest --verbose --showlocals -s functional_test_examples/ --installdir=[syslog-ng install path]`
 * Example2: Start only special testcase (or testcases) which are matching on "pattern from testcase"
   * `$ cd tests/pytest_framework`
   * `$ python3 -m pytest --verbose --showlocals -s functional_test_examples/ -k [pattern from testcase] --installdir=[syslog-ng install path]`

4, Reporting:
 * The framework uses capabilities of the builtin `logging` module.
 * Available log levels: INFO (default), DEBUG, ERROR
 * Console log:
   * Pytest display verbose log from testcases when it was started with: "--verbose --showlocals -s"
   * Without this start arguments pytest will not log anything on console, only if there is an error in testcase
 * File log:
   * Self unit-tests are not, but functional-tests are using a separate report directory which contains followings:
     * syslog-ng related files: config, pid, persist, ctl, console stdout/stderr 
     * source-destination related files: input/output files
     * testcase related files: full testcase log, also a started testcase

## The minimal feature set is the following: 
- user can generate config with a file source and file destination
    - mandatory options are not required, framework can be generate it automatically
    - multiple file sources and file destinations can be used (1 statement uses 1 file driver)
- user can start/reload/stop syslog-ng
    - wrong configuration can be detected
    - separate syslog-ng instances can be started with different configurations
- user can generate BSD message(s)
    - message fields can be filled by user or use defaults
- user can write message(s) into the resource of the file source
- user can read message(s) from the resource of the file destination
- expected message can be generated from written message
- expected and read messages can be compared (assert)
- every testcase is logging into a separate directory (reports/)
    - tescase log file is saved
    - syslog-ng session files are saved

## Additional features which are missing from this version:
- using raw configuration
- update driver options after syslog-ng start
- core file detection
- run syslog-ng behind other tools: strace, valgrind
- generate IETF message
- get and compare (assert) stats/query counters from syslog-ng-ctl

### CODE REVIEW

**Comment by kira-syslogng** [comment]
success

**Comment by mitzkia** [comment]
skip ci

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by mitzkia** [comment]
@kira-syslogng retest this please

**Comment by furiel** [comment]
@kira-syslogng retest this please

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
failure

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by mitzkia** [comment]
@kira-syslogng retest this please

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by mitzkia** [comment]
@kira-syslogng retest this please

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by mitzkia** [comment]
@kira-syslogng retest this please

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by mitzkia** [comment]
@kira-syslogng retest this please test branch=micek_java_destination_crash; test=functions/destination-drivers/java-destination-driver/013-java-destination-group-remove-add.zts;

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by mitzkia** [comment]
@kira-syslogng retest this please

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by mitzkia** [comment]
This PR is failing on Kira because.
- PR contains a patch which added requirements (pip) packages to the DBLDs package list
With this patch we can use later new make targets (make pytest-check) in DBLD images.
- the current Xenial image in DBLD (which Kira builds) is structured in wrong way: pip package installs are before apt installs. There is a dependency in pip install (gcc package needed to install psutil package) which can be installed only after pip packages.

So for now this PR is failing, but after #2318 is merged (which also fixes the wrong install precedency) I can rebase to the master to fix this issue.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

File: tests/pytest_framework/src/syslog_ng_ctl/syslog_ng_ctl_executor.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        self.command_executor = CommandExecutor(logger_factory)
+
+    def run_command(self, command_short_name, command):
+        return self.command_executor.run(
+            command=self.construct_ctl_command(command),
+            stdout_path=self.construct_std_file_path(command_short_name, "stdout"),
+            stderr_path=self.construct_std_file_path(command_short_name, "stderr")
+        )
+
+    def construct_std_file_path(self, command_short_name, std_type):
```
Comment (MrAnno) [inline tests/pytest_framework/src/syslog_ng_ctl/syslog_ng_ctl_executor.py]: I think this should be private.

File: tests/pytest_framework/src/syslog_ng_ctl/syslog_ng_ctl_executor.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        instance_name = self.instance_paths.get_instance_name()
+        return Path(self.working_dir, "syslog_ng_ctl_{}_{}_{}".format(instance_name, command_short_name, std_type))
+
+    def construct_ctl_command(self, command):
```
Comment (MrAnno) [inline tests/pytest_framework/src/syslog_ng_ctl/syslog_ng_ctl_executor.py]: This should be private as well.

File: tests/pytest_framework/src/syslog_ng_ctl/syslog_ng_ctl.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        return self.syslog_ng_ctl_as_command.reload()
+
+    def stop(self):
+        return self.syslog_ng_ctl_as_command.stop()
+
+    def wait_for_control_socket_alive(self):
+        return self.syslog_ng_ctl_as_command.wait_for_control_socket_alive()
+
+    def wait_for_control_socket_stopped(self):
+        return self.syslog_ng_ctl_as_command.wait_for_control_socket_stopped()
```
Comment (MrAnno) [inline tests/pytest_framework/src/syslog_ng_ctl/syslog_ng_ctl.py]: I would just merge this class with `SyslogNgCtlAsCommand` and `SyslogNgCtlExecutor`.
This looks like a wrapper class without any further value.

File: tests/pytest_framework/src/syslog_ng/syslog_ng.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    def reload(self, config):
+        self.__syslog_ng_cli.reload(config)
+
+    def restart(self, config):
+        self.__syslog_ng_cli.start(config)
+        self.__syslog_ng_cli.stop()
+
+    def get_version(self):
+        return self.__syslog_ng_cli.get_version()
```
Comment (MrAnno) [inline tests/pytest_framework/src/syslog_ng/syslog_ng.py]: This class just wraps a `SyslogNgCli` instance without modifying its interface or adding anything else.
Are you planning to add state or different methods to this class? If not, I would just merge `SyslogNgCli` into `SyslogNg`.

**Comment by MrAnno** [comment]
Your commit messages are really good, they were easy and straightforward to review. Thanks for separating your changes! :+1:

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by mitzkia** [comment]
@MrAnno Thanks for the review notes. I fixed some of them.
And added some other fixes. From my side it would not change else.

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by bazsi** [comment]
\o/

**Pull request merged**